File: InterviewPrepKit/Arrays/NewYearChaos/mergesort.py
def sort(ary, lb, mid, ub):
    # pivot on left partition first
    i_inspect = lb
    i_right_partition = mid + 1 # note that i_right_partition will advance

    inv_count = 0
    
    while i_inspect < mid+1:
        if ary[i_inspect] > ary[i_right_partition]:	# then item in left partition is greater than item in right partition
            # there has been an inversion
            inv_count += mid-i_inspect + 1

            # so swap the value we are inspecting in the left partition with value we are comparing it to in the right partition
            val_left_partition = ary[i_inspect]
            ary[i_inspect] = ary[i_right_partition]
            ary[i_right_partition] = val_left_partition	
        
        if i_right_partition < ub:
            i_right_partition += 1 # advance the right-comparison index to the next element
        else:   # then we have compared the pivot ("inspection") element in the left partition to every element in the right partition
            i_inspect += 1  # so we can now advance to the next pivot ("inspection") element in the left partition
            i_right_partition = mid + 1 # now we need to reset the right-position "pointer" back to the beginning (of the right partition)
        

    # now reconcile (sort) right partition
    i_inspect = mid + 1 # obviously we start inspection at first elt in right partition
    i_right_partition = mid + 2 # and we begin with the second element in the right partition (since it is pointless to compared the inspection element to itself)
    while i_inspect < ub:
        if ary[i_inspect] > ary[i_right_partition]:
            # so swap the value we are inspecting in the right partition with value we are comparing it to (also in the right partition)
            val_right_partition = ary[i_inspect]
            ary[i_inspect] = ary[i_right_partition]
            ary[i_right_partition] = val_right_partition	
            
        if i_right_partition < ub:
            i_right_partition += 1
        else:
            i_inspect += 1
            i_right_partition = i_inspect+1

    return inv_count
			
		
# merge_sort uses the concept known as Divide and Conquer
def merge_sort(ary, lb, rb):
    # this is a recursive function, so we need a base case
    if lb == rb:	# then we have sub-divided ary until the point that the current sub-array contains only a single element
        return 0

    inv_count = 0
        
    # otherwise, sub-divide ary,	
    # e.g. {74, 4, -12, 8, 9, 7, 2, 0}		should produce the following logical subdivisions
    # 		  |				 |
    # {-74, 4, -12, 8}	{9, 7, 2, 0}
    # 	  |   	   |	  |		  |
    # {-74, 4} {-12, 8}	 {9, 7}   {2, 0}
    #   |    |   |    |   |   |    |  |
    # {-74} {4} {-12} {8} {9} {7} {2} {0}
    
    # first step is to find the index that will sub-divide ary into two logical partitions
    mid = (lb + rb) // 2	# note that since this is integer division, we will implicitly round down
    
    # now recursively partition ary into left and right sub-arrays
    # left partition will consist of elements from ary ranging from indexes lb to mid
    inv_count += merge_sort(ary, lb, mid)
    # right partition will consist of elements from ary ranging from indexes mid+1 to rb
    inv_count += merge_sort(ary, mid+1, rb)
    
    # now we need to sort/merge the left and right partitions
    inv_count += sort(ary, lb, mid, rb)

    return inv_count


def mergesort(ary):
    return merge_sort(ary, 0, len(ary)-1)

File: InterviewPrepKit/Arrays/NewYearChaos/MergeSort.java
import java.util.Scanner;

public class MergeSort {

   /* Define your method here */
   
   // NOTE to instructor:
   //    This is from an old assignment of mine from my own implementation of MergeSort
   //    This is my own code and is at least two years old.
   
   private static void sortMerge(int[] ary, int lbound, int leftpartEnd, int rbound) {
		//pivot on left partition first
		int inspectPos = lbound;
		int rightpartPos = leftpartEnd+1;
		
		while (inspectPos < leftpartEnd+1) {
			if (ary[inspectPos] > ary[rightpartPos]) {	//then item in left partition is greater than item in right partition
				//so swap items
				int swap = ary[inspectPos];
				ary[inspectPos] = ary[rightpartPos];
				ary[rightpartPos] = swap;	
			}
			if (rightpartPos < rbound)
				rightpartPos++;
			else {
				inspectPos++;
				rightpartPos = leftpartEnd+1;
			}
		}
		
		//now reconcile right partition
		inspectPos = leftpartEnd+1;
		rightpartPos = leftpartEnd+2;
		while (inspectPos < rbound) {
			if (ary[inspectPos] > ary[rightpartPos]) {
				//so swap items
				int swap = ary[inspectPos];
				ary[inspectPos] = ary[rightpartPos];
				ary[rightpartPos] = swap;	
			} 
			if (rightpartPos < rbound)
				rightpartPos++;
			else {
				inspectPos++;
				rightpartPos = inspectPos+1;
			}
		}
	}
	
	//this function applies Divide and Conquer
	private static void mergeSort(int[] ary, int lbound, int rbound) {
		//this is a recursive function, so we need a base case
		if (lbound == rbound)	//then we have sub-divided ary until the point that the current sub-array contains only a single element
			return;
			
		//otherwise, sub-divide ary,	
		//e.g. {74, 4, -12, 8, 9, 7, 2, 0}		should produce the following logical subdivisions
		//		  |				 |
		//{-74, 4, -12, 8}	{9, 7, 2, 0}
		//	  |   	   |	  |		  |
		//{-74, 4} {-12, 8}	 {9, 7}   {2, 0}
		//  |    |   |    |   |   |    |  |
		//{-74} {4} {-12} {8} {9} {7} {2} {0}
		
		//first step is to find the index that will sub-divide ary into two logical partitions
		int lpartEnd = (lbound + rbound)/2;	//note that since this is an int, we will implicitly round down
		
		//now recursively partition ary into left and right sub-arrays
		//left partition will consist of elements from ary ranging from indexes lbound to lpartEnd
		mergeSort(ary, lbound, lpartEnd);
		//right partition will consist of elements from ary ranging from indexes lpartEnd+1 to rbound
		mergeSort(ary, lpartEnd+1, rbound);
		
		//now we need to sort/merge the left and right partitions
		sortMerge(ary, lbound, lpartEnd, rbound);
	}
	
	public static void sortArray(int[] myArr, int arrSize) {
	   mergeSort(myArr, 0, arrSize-1);
	}


   public static void main(String[] args) {
      /* Type your code here. */
         Scanner scnr = new Scanner(System.in);
         int n_ints = scnr.nextInt();
         int[] i_arr = new int[n_ints];
         
         for (int i = 0; i < n_ints; i++) {
            i_arr[i] = scnr.nextInt();
         }
         
         sortArray(i_arr, i_arr.length);
         
         for (int i = 0; i < n_ints; i++) {
            System.out.printf("%d ", i_arr[i]);
         }
         System.out.println();
   }
}